content/blog/2019/11/2019-11-21-welcome-to-the-matrix.adoc

```diff
@@ -0,0 +1,389 @@
+
+.Jenkinsfile
+[source, groovy]
+----
+pipeline {
+    agent none
+    stages {
+        stage('BuildAndTest') {
+            matrix {
+                agent any
+                axes {
+                    axis {
+                        name 'PLATFORM'
+                        values 'linux', 'windows', 'mac'
+                    }
+                    axis {
+                        name 'BROWSER'
+                        values 'firefox', 'chrome', 'safari', 'edge'
+                    }
+                }
+                stages {
+                    stage('Build') {
+                        steps {
+                            echo "Do Build for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                    stage('Test') {
+                        steps {
+                            echo "Do Test for ${PLATFORM} - ${BROWSER}"
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+----
+
+.Log output (truncated)
+[source]
+----
+...
+[Pipeline] stage
+[Pipeline] { (BuildAndTest)
+[Pipeline] parallel
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'firefox')
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'chrome')
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'safari')
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'safari')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'safari')
+[Pipeline] { (Branch: Matrix - OS = 'linux', BROWSER = 'edge') (hide)
+[Pipeline] { (Branch: Matrix - OS = 'windows', BROWSER = 'edge')
+[Pipeline] { (Branch: Matrix - OS = 'mac', BROWSER = 'edge')
+...
+Do Build for linux - safari
+Do Build for linux - firefox
+Do Build for windows - firefox
+Do Test for linux - firefox
+Do Build for mac - firefox
+Do Build for linux - chrome
+Do Test for windows - firefox
+...
+----
+
+
+== Excluding invalid combinations
+
+Now that I have my basic matrix created, I've noticed that I have some invalid combinations.  
+Microsoft Edge only runs on Windows and there isn't a Linux version of Safari. 
+
+I can remove invalid cells from my matrix using `exclude` directives. Each `exclude` has one or more `axis` directives with `name` and `values`.  
+The `axis` directives inside an `exclude` generate a set of combinations (similar to generating the matrix cells). 
+The matrix cells that match all the values from an `exclude` combination are removed from the matrix. 
+If I have more than one `exclude` directive, each are evaluated separately to remove cells.
+
+When dealing with a long lists of values to exclude, I can use a `notValues` instead of `values` to specify axis values we *don't* want excluded.
```

Comment:
```suggestion
When dealing with a long lists of values to exclude, I can use `notValues` instead of `values` to specify axis values we *don't* want excluded.
```